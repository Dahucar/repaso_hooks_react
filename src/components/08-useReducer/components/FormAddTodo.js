import React from 'react'
import { useForm } from '../../../hooks/useForm';

export const FormAddTodo = ({ handleAddTodo }) => {
    const [ { desc }, handleInputChangue, resetInputsValues ] = useForm({
        desc: ''
    });
    const handleSubmit = ( e ) => {
        e.preventDefault();
        // validacion basica de entrada. 
        if (desc.trim().length <= 1) {
            return;
        }        
        // creando la nueva todo / tarea
        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }
        // enviamos la accion mediante el dispatch al reducer
        handleAddTodo( newTodo );
        resetInputsValues();
    }
    return (
        <>
            <h3 className="text-center">Agregar todo</h3>
            <hr />
            <form className="mr-4">
                <input 
                    type="text"
                    name="desc" 
                    className="form-control"
                    placeholder="Ingrese tarea" 
                    autoComplete="off" 
                    value={ desc }
                    onChange={ handleInputChangue }
                />
                <button
                    type="submit"
                    className="btn btn-outline-success mt-1 btn-block"
                    onClick={ handleSubmit }
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
