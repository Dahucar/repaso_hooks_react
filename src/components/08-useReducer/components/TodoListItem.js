import React from 'react'

export const TodoListItem = ({ todo, index, hanldeToggleTodo, handleDelete }) => {
    return (
        <li className="list-group-item">
            <p className={ `${ todo.done && 'complete' }` }
                onClick={ () => hanldeToggleTodo( todo.id ) }    
            >
                <span className={ `mr-2 badge badge-pill ${ todo.done ? 'badge-success' : 'badge-danger' }` }>
                    { index + 1 }
                </span>
                { todo.desc }
            </p>
            <button className="btn btn-danger btn-sm"
                onClick={ () => handleDelete( todo.id ) }>
                Quitar
            </button>
        </li>
    )
}
