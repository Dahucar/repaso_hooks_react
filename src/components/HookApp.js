import React from 'react'
import { useCustomCounter } from './01-useState/useCustomCounter'

export const HookApp = () => {

    const { state, increment, decrement, defaultValue } = useCustomCounter();

    return (
        <>
            <h3> {state} </h3>
            <button className="btn btn-success"
                onClick={() => increment(2)}>
                Increment
            </button>
            <button className="btn btn-primary"
                onClick={() => decrement(4)}>
                Decrement
            </button>
            <button className="btn btn-danger"
                onClick={defaultValue}>
                Default value
            </button>
        </>
    )
}
