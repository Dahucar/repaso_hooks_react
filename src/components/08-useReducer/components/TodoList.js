import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, hanldeToggleTodo, handleDelete }) => {
    return (
        <div className="card">
            {/* <TodoList todos, hanldeToggleTodo, handleDelete /> */}
            <ul className="list-group list-group-flush">
                {/* <TodoListItem todoItem, index, hanldeToggleTodo, handleDelete/> */}
                {
                    todos.map( (todo, i) => (
                        <TodoListItem 
                            key={ todo.id }
                            todo={ todo } 
                            index={ i } 
                            hanldeToggleTodo={ hanldeToggleTodo }
                            handleDelete={ handleDelete }
                        />
                    ))
                }
            </ul>
        </div>
    )
}
