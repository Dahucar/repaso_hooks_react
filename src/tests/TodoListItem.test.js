import { shallow } from "enzyme";
import { TodoListItem } from "../components/08-useReducer/components/TodoListItem";
import { demoTodos } from "../data/demoTodo";

describe('Pruebas en el <TodoListItem />', () => {

    const hanldeToggleTodo = jest.fn();
    const handleDelete = jest.fn(); 

    const wrapper = shallow( 
        <TodoListItem 
            todo={ demoTodos[0] } 
            index={ demoTodos[0].id } 
            hanldeToggleTodo={ hanldeToggleTodo }
            handleDelete={ handleDelete }
        /> 
    );
    test('Debe hacer match con la instantanea', () => {
        expect( wrapper ).toMatchSnapshot();   
    });
    test('Debe de llamar la funcion handleDelete', () => {
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenLastCalledWith( demoTodos[0].id );
    });
    test('Debe de llamar la funcion hanldeToggleTodo', () => {
        wrapper.find('p').simulate('click');
        expect( hanldeToggleTodo ).toHaveBeenLastCalledWith( demoTodos[0].id ); 
    });
    test('Debe de mostrar el texto correctamente <p>', () => {
        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe((demoTodos[0].id + 1)+demoTodos[0].desc);
    });
});
