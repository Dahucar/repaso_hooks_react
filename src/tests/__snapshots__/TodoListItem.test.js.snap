// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <TodoListItem /> Debe hacer match con la instantanea 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    <span
      className="mr-2 badge badge-pill badge-danger"
    >
      2
    </span>
    Aprender React
  </p>
  <button
    className="btn btn-danger btn-sm"
    onClick={[Function]}
  >
    Quitar
  </button>
</li>
`;
